### DISCUSSION TITLE
Support strictmode in Typescript target

### DISCUSSION THREAD

**Discussion by ChrisLincoln** [antlr/antlr4#4090] [General]
I've modified the Dockerfile to be able to use the head of the dev branch so I can try out the new Typescript target. I have a TS project with strict mode turned on. The generated lexer.ts and parser.ts contain code that will not compile in strict mode.  For example, using the sample [arithmetic](https://github.com/antlr/grammars-v4/blob/master/arithmetic/arithmetic.g4) grammer, you get these:

```
	public static readonly literalNames: string[] = [ null, null, null, "'('", 
                                                   "')'", "'+'", "'-'", 
                                                   "'*'", "'/'", "'>'", 
                                                   "'<'", "'='", "'.'", 
                                                   "'^'" ];
```

To pass in strict mode, these should be (string|null)[].  Also, the generated code is passing class types in the calls to getTypedRuleContext(), which seems correct per the underlying runtime javascript, but is throwing strict mode errors:

```
	public equation_list(): EquationContext[] {
		return this.getTypedRuleContexts(EquationContext) as EquationContext[];
	}
```

```
arithmeticParser.ts:486:36 - error TS2345: Argument of type 'typeof EquationContext' is not assignable to parameter of type 'new (parser?: Parser | undefined, parent?: ParserRuleContext | undefined, invokingState?: number | undefined, ...args: any[]) => EquationContext'.
  Types of construct signatures are incompatible.
    Type 'new (parser?: arithmeticParser | undefined, parent?: ParserRuleContext | undefined, invokingState?: number | undefined) => EquationContext' is not assignable to type 'new (parser?: Parser | undefined, parent?: ParserRuleContext | undefined, invokingState?: number | undefined, ...args: any[]) => EquationContext'.
      Types of parameters 'parser' and 'parser' are incompatible.
        Type 'Parser | undefined' is not assignable to type 'arithmeticParser | undefined'.
          Type 'Parser' is missing the following properties from type 'arithmeticParser': grammarFileName, literalNames, symbolicNames, ruleNames, and 11 more.

486   return this.getTypedRuleContexts(EquationContext) as EquationContext[];

```

I don't really know how to fix those. If I could turn off strict mode for just these files, I would and not worry about it because I inherently trust in their goodness as auto-generated code. But I can't turn it off.  I suppose I could spin off a "shared" project and turn off strict mode in its compilation.

**Comment by ericvergnaud** [comment]
So you want to explicitly turn strict mode off for the generated files ?Envoyé de mon iPhoneLe 29 janv. 2023 à 16:41, ChrisLincoln ***@***.***> a écrit :﻿
I've modified the Dockerfile to be able to use the head of the dev branch so I can try out the new Typescript target. I have a TS project with strict mode turned on. The generated lexer.ts and parser.ts contain code that will not compile in strict mode.  For example, using the sample arithmetic grammer, you get these:
	public static readonly literalNames: string[] = [ null, null, null, "'('", 
                                                   "')'", "'+'", "'-'", 
                                                   "'*'", "'/'", "'>'", 
                                                   "'<'", "'='", "'.'", 
                                                   "'^'" ];

To pass in strict mode, these should be (string|null)[].  Also, the generated code is passing class types in the calls to getTypedRuleContext(), which seems correct per the underlying runtime javascript, but is throwing strict mode errors:
	public equation_list(): EquationContext[] {
		return this.getTypedRuleContexts(EquationContext) as EquationContext[];
	}

arithmeticParser.ts:486:36 - error TS2345: Argument of type 'typeof EquationContext' is not assignable to parameter of type 'new (parser?: Parser | undefined, parent?: ParserRuleContext | undefined, invokingState?: number | undefined, ...args: any[]) => EquationContext'.
  Types of construct signatures are incompatible.
    Type 'new (parser?: arithmeticParser | undefined, parent?: ParserRuleContext | undefined, invokingState?: number | undefined) => EquationContext' is not assignable to type 'new (parser?: Parser | undefined, parent?: ParserRuleContext | undefined, invokingState?: number | undefined, ...args: any[]) => EquationContext'.
      Types of parameters 'parser' and 'parser' are incompatible.
        Type 'Parser | undefined' is not assignable to type 'arithmeticParser | undefined'.
          Type 'Parser' is missing the following properties from type 'arithmeticParser': grammarFileName, literalNames, symbolicNames, ruleNames, and 11 more.

486   return this.getTypedRuleContexts(EquationContext) as EquationContext[];


I don't really know how to fix those. If I could turn off strict mode for just these files, I would and not worry about it because I inherently trust in their goodness as auto-generated code. But I can't turn it off.  I suppose I could spin off a "shared" project and turn off strict mode in its compilation.

**Reply by ChrisLincoln** [reply]
Just to be clear, I'm referring to Typescript 'strict' mode, and not JS 'use strict'. TS doesn't support {strict: false} at the file or sub-folder level, as far as I can tell.  For the moment, I've turned strict mode off for the project so I can move on.  What I am battling now is just basic ESM module stuff with ts-node.

Do you know of a minimal node project that demonstrates use of this runtime?  I'd like to see what the package.json and tsconfig.json contents are.

**Reply by ericvergnaud** [reply]
Yes I also believe that strict mode at file level is not possible in TypeScript
You can check https://github.com/compose-lang/compiler
It doesn't use strict mode

**Reply by ChrisLincoln** [reply]
Thank you for the sample reference. It confirms the changes that I was able to get to work just a moment ago--which I lifted from the tsconfig and package.json from the main runtime repo--using a trivial "hello antlr" node project.  The task ahead of me is to apply the changes to a much larger project that has some not-friendly-to-esm code. It's a long overdue task.

**Comment by ericvergnaud** [comment]
beta 10 brings some minor tweaks
I would need the generated parser in order to understand the parser error with EquationContext
